lib/cloudkeeper/one/appliance_actions/removal.rb

```diff
@@ -3,31 +3,52 @@ module One
     module ApplianceActions
       module Removal
         def remove_appliance(appliance_id)
+          logger.debug "Removing templates for appliance #{appliance_id.inspect}"
           remove_templates :find_by_appliance_id, appliance_id
+          logger.debug "Removing images for appliance #{appliance_id.inspect}"
           remove_images :find_by_appliance_id, appliance_id
         end
 
         def remove_image_list(image_list_id)
+          logger.debug "Removing templates for image list #{image_list_id.inspect}"
           remove_templates :find_by_image_list_id, image_list_id
+          logger.debug "Removing images for image list #{image_list_id.inspect}"
           remove_images :find_by_image_list_id, image_list_id
         end
 
         def remove_expired
-          image_handler.expired.each { |image| image_handler.delete image }
+          logger.debug 'Removing expired images...'
+          handle_iteration(image_handler.expired) { |item| image_handler.delete item }
         end
 
         def remove_templates(method, value)
-          templates = template_handler.send(method, value)
-          templates.each { |template| template_handler.delete template }
+          handle_iteration(template_handler.send(method, value)) { |item| template_handler.delete item }
         end
 
         def remove_images(method, value)
-          images = image_handler.send(method, value)
-          images.each do |image|
-            image_handler.expire image
-            image_handler.delete image
+          handle_iteration(image_handler.send(method, value)) do |item|
+            image_handler.expire item
+            image_handler.delete item
           end
         end
+
+        private
+
+        def handle_iteration(items)
+          raise Cloudkeeper::One::Errors::ArgumentError, 'Iteration error handler was called without a block!' unless block_given?
+
+          error = nil
+          items.each do |item|
+            begin
+              yield item
+            rescue Cloudkeeper::One::Errors::StandardError => ex
+              error = ex
+              next
+            end
+          end
+
+          raise error if error
```

Comment:
So, in case of multiple errors (on different items), you remember only the last one? That seems a bit strange to me. Especially since you are re-raising the last error.